torchdata/dataloader2/dataloader2.py

```diff
@@ -43,6 +43,38 @@ class ConcurrencySpec:
     persistent_workers: bool = False
 
 
+class _DataLoader2Iterator(Iterator):
+    def __init__(self, dataloader: "DataLoader2", iterator_id: int):
+        self.dataloader = dataloader
+        self.iterator_id = iterator_id
+
+    def __next__(self) -> T_co:
+        if self.iterator_id == self.dataloader.valid_iterator_id:
+            self.dataloader._reset_iter = True
+            try:
+                return next(self.dataloader._datapipe_iter)  # type: ignore[arg-type]
+            except PauseIteration:
+                raise StopIteration
+            except StopIteration:
+                if self.dataloader.reading_service is not None:
+                    self.dataloader.reading_service.finalize_iteration()
+                raise
+        else:
+            raise RuntimeError(
+                "This iterator has been invalidated because another iterator has been created "
+                "from the same DataLoader2.\n"
+                "This may be caused multiple references to the same DataLoader2. "
+                "For feedback regarding this single iterator per DataLoader2 constraint, feel free "
+                "to comment on this issue: https://github.com/pytorch/data/issues/45."
+            )
+
+    def __getattr__(self, name):
+        """
+        To delegate operations such as `limit`, `pause`, `resume`, `__getstate__`
```

Comment:
And, please remove examples of `limit` and `pause` for now because those APIs are tied to internal services